primitives/src/subset.rs

```diff
@@ -44,30 +45,54 @@ pub trait LatticeOrd {
 /// considered to be bigger than any finite set of its elements. This is true for infinite
 /// sets. When talking about finite sets, we have to add that they are _open_.
-///
-/// The meaning of the parameter:
-/// - `None`: no restrictions, the whole set.
-/// - `Some(elems)`: the subset of elements in `elems`.
-///    A particular case is the empty set where `elems` is empty.
-#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Default)]
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
-pub struct SubsetRestriction<A: Ord>(pub Option<BTreeSet<A>>);
+pub enum SubsetRestriction<A: Ord> {
+    /// No restrictions, the whole set.
+    Whole,
+    /// Exactly these elements, and no others.
+    These(BTreeSet<A>),
+    /// The whole set except these elements.
+    Except(BTreeSet<A>),
+}
+
+impl<A: Ord> Default for SubsetRestriction<A> {
+    fn default() -> Self {
+        Self::Whole
+    }
+}
 
 impl<A> LatticeOrd for SubsetRestriction<A>
 where
     A: Clone + Ord + PartialEq,
 {
     fn lattice_cmp(&self, other: &Self) -> LatticeOrdering {
-        match (&self.0, &other.0) {
-            (None, None) => LatticeOrdering::Equal,
-            (_, None) => LatticeOrdering::Less,
-            (None, _) => LatticeOrdering::Greater,
-            (Some(a), Some(b)) => match (a.is_subset(b), b.is_subset(a)) {
-                (true, true) => LatticeOrdering::Equal,
-                (true, false) => LatticeOrdering::Less,
-                (false, true) => LatticeOrdering::Greater,
-                _ => LatticeOrdering::Incomparable,
-            },
+        let left = match self {
+            Self::Except(es) if es.is_empty() => &Self::Whole,
+            x => x,
+        };
+        let right = match other {
+            Self::Except(es) if es.is_empty() => &Self::Whole,
+            x => x,
+        };
+        let cmp_same = |a: &BTreeSet<_>, b: &BTreeSet<_>| match (a.is_subset(b), b.is_subset(a)) {
+            (true, true) => LatticeOrdering::Equal,
+            (true, false) => LatticeOrdering::Less,
+            (false, true) => LatticeOrdering::Greater,
+            _ => LatticeOrdering::Incomparable,
+        };
+        let cmp_diff = |a: &BTreeSet<_>, b: &BTreeSet<_>| match a.intersection(b).next() {
+            Some(_) => LatticeOrdering::Incomparable,
+            None => LatticeOrdering::Less,
+        };
+        match (left, right) {
+            (Self::Whole, Self::Whole) => LatticeOrdering::Equal,
+            (_, Self::Whole) => LatticeOrdering::Less,
+            (Self::Whole, _) => LatticeOrdering::Greater,
+            (Self::These(a), Self::These(b)) => cmp_same(a, b),
+            (Self::Except(a), Self::Except(b)) => cmp_same(b, a),
+            (Self::These(a), Self::Except(b)) => cmp_diff(a, b),
+            (Self::Except(a), Self::These(b)) => cmp_diff(b, a),
```

Comment:
Ordering of a and b doesn't matter since we are using the intersection. Using reverse order here only adds a misdirection in understanding the code.